// @astrojs/mdx/dist/satteri/index.js
import { pathToFileURL } from "node:url";
import {
  satteriCollectImagesPlugin,
  satteriCreateHighlightFn,
  satteriHeadingIdsPlugin,
  satteriHighlightPlugin
} from "@astrojs/markdown-satteri";
import { createDefaultAstroMetadata } from "astro/markdown";
import {
  mdxToJs
} from "satteri";
import { ASTRO_IMAGE_IMPORT, USES_ASTRO_IMAGE_FLAG } from "../image-constants.js";
import { shouldAddCharset } from "./charset.js";
import { createAstroMetadataPlugin } from "./hast-astro-metadata.js";
import { createImageToComponentPlugin } from "./hast-images-to-component.js";
function createMdxProcessor(mdxOptions, satteriOptions, ctx) {
  let highlightFn;
  let initPromise;
  function initHighlighter() {
    initPromise = satteriCreateHighlightFn(mdxOptions.syntaxHighlight, mdxOptions.shikiConfig).then(
      (fn) => {
        highlightFn = fn;
      }
    );
  }
  return {
    async process(content, filePath, frontmatter) {
      if (!highlightFn && !initPromise) {
        initHighlighter();
      }
      if (initPromise) await initPromise;
      const headings = [];
      const localImagePaths = /* @__PURE__ */ new Set();
      const remoteImagePaths = /* @__PURE__ */ new Set();
      const astroMetadata = createDefaultAstroMetadata();
      const collectImages = satteriCollectImagesPlugin(localImagePaths, remoteImagePaths);
      const headingIds = satteriHeadingIdsPlugin(headings, frontmatter);
      const astroMeta = createAstroMetadataPlugin(astroMetadata, filePath);
      const imageImportInfo = {
        importedImages: /* @__PURE__ */ new Map(),
        hasImages: false
      };
      const imageToComponent = createImageToComponentPlugin(
        localImagePaths,
        remoteImagePaths,
        imageImportInfo
      );
      const syntaxHighlight = mdxOptions.syntaxHighlight;
      const excludeLangs = typeof syntaxHighlight === "object" ? syntaxHighlight.excludeLangs : void 0;
      const allMdastPlugins = satteriOptions.mdastPlugins.length ? [collectImages, ...satteriOptions.mdastPlugins] : [collectImages];
      const hastPlugins = [];
      if (highlightFn) {
        hastPlugins.push(satteriHighlightPlugin(highlightFn, excludeLangs, { mdx: true }));
      }
      if (satteriOptions.hastPlugins.length) {
        hastPlugins.push(...satteriOptions.hastPlugins);
      }
      hastPlugins.push(imageToComponent, headingIds, astroMeta);
      let optimizeStatic;
      if (mdxOptions.optimize) {
        const ignoreElements = typeof mdxOptions.optimize === "object" ? mdxOptions.optimize.ignoreElementNames : void 0;
        optimizeStatic = {
          component: "Fragment",
          prop: "set:html",
          ...ignoreElements && { ignoreElements }
        };
      }
      const mdxResult = await mdxToJs(content, {
        mdastPlugins: allMdastPlugins,
        hastPlugins,
        optimizeStatic,
        features: {
          ...satteriOptions.features,
          // `mdxOptions.gfm`/`smartypants` are always boolean-shaped; skip the override when
          // satteri's feature is an object so granular config isn't clobbered.
          ...typeof satteriOptions.features.gfm === "object" ? {} : { gfm: mdxOptions.gfm !== false },
          ...typeof satteriOptions.features.smartPunctuation === "object" ? {} : { smartPunctuation: mdxOptions.smartypants !== false }
        },
        fileURL: pathToFileURL(filePath),
        jsxImportSource: "astro"
      });
      let compiled = mdxResult.code;
      compiled = compiled.replace(/^export default MDXContent;\s*$/m, "");
      if (imageImportInfo.hasImages) {
        compiled += `
import { Image as ${ASTRO_IMAGE_IMPORT} } from "astro:assets";`;
        for (const [src, importName] of imageImportInfo.importedImages) {
          compiled += `
import ${importName} from ${JSON.stringify(src)};`;
        }
        compiled += `
export const ${USES_ASTRO_IMAGE_FLAG} = true;`;
      }
      compiled += `
export const frontmatter = ${JSON.stringify(frontmatter)};`;
      compiled += `
export function getHeadings() { return ${JSON.stringify(headings)}; }`;
      if (frontmatter.layout) {
        compiled = compiled.replace(/^function MDXContent\(/m, "function __OriginalMDXContent__(");
        compiled += `
import { jsx as __astro_layout_jsx__ } from 'astro/jsx-runtime';
import __astro_layout_component__ from ${JSON.stringify(frontmatter.layout)};
export default function MDXContent(props) {
	const content = __OriginalMDXContent__(props);
	const { layout, ...frontmatterContent } = frontmatter;
	frontmatterContent.file = file;
	frontmatterContent.url = url;
	return __astro_layout_jsx__(__astro_layout_component__, {
		file,
		url,
		content: frontmatterContent,
		frontmatter: frontmatterContent,
		headings: getHeadings(),
		'server:root': true,
		children: content,
	});
}`;
      } else if (shouldAddCharset(content, filePath, ctx.srcDir)) {
        compiled = compiled.replace(/^function MDXContent\(/m, "function __OriginalMDXContent__(");
        compiled += `
import { jsx as __astro_charset_jsx__, jsxs as __astro_charset_jsxs__, Fragment as __astro_charset_Fragment__ } from 'astro/jsx-runtime';
export default function MDXContent(props) {
	return __astro_charset_jsxs__(__astro_charset_Fragment__, {
		children: [
			__astro_charset_jsx__('meta', { charset: 'utf-8' }),
			__OriginalMDXContent__(props),
		],
	});
}`;
      }
      return {
        code: compiled,
        astroMetadata
      };
    }
  };
}
export {
  createMdxProcessor
};
